### PR TITLE
Added "tags" to brushes and filtering based on them.

### PR DESCRIPTION
**Pull request by insominx**
Currently the brush panel is cluttered with a large number of brushes and is likely to grow larger in the future.  As a first step toward better organizing them, this PR serves to classify each brush with a list of tags and a couple of filtering options.

Via OpenBrush.cfg (UserConfig), the brushes that show up in the panel can be limited by using the following options:
1) IncludeTags - specifies all of the brushes that may be included included in the panel
2) ExcludeTags - specifies all of the brush that must be excluded (even if they are also in "include").

To test, try the following:
- Base case
 `{
  "User": {
  },
  "Brushes": {
  },
  "Video": {
  },
  "Flags": {
  },
  "Export": {
  },
}`
Result should be the same as before this PR.
- Declare include  
`{
  "User": {
  },
  "Brushes": {
    "IncludeTags": ["classroom"],
  },
  "Video": {
  },
  "Flags": {
  },
  "Export": {
  },
}
`
Result should be only a single page of brushes that have been tagged with "classroom".
- Add a new tag called "test" to a few brush descriptors that also have "classroom" and then declare inlude and exclude
`{
  "User": {
  },
  "Brushes": {
    "IncludeTags": ["classroom"],
    "ExcludeTags": ["test"]
  },
  "Video": {
  },
  "Flags": {
  },
  "Export": {
  },
}`
Result should be only a single page of brushes that have been tagged with "classroom" minus the brush tagged with "test".
- Exclude only case
`{
  "User": {
  },
  "Brushes": {    
    "ExcludeTags": ["classroom"]
  },
  "Video": {
  },
  "Flags": {
  },
  "Export": {
  },
}`
Result should be all of the brushes minus those tagged with classroom.

In addition, tags can be added and removed.  Here is an example that includes this usage:
```
{
	"User": {
	},
	"Brushes": {
		"AddTagsToBrushes": {
			"Rainbow": ["classroom", "bedazzling"],
			"Plasma": ["classroom", "bedazzling"],
			"testBrushNotFound": ["clasroom"]
		},
		"RemoveTagsFromBrushes": {
			"Rainbow": ["default", "test"],
			"Plasma": ["default", "test"],
			"testBrushNotFound": ["clasroom"]
		},
		"IncludeTags": ["classroom"],
		"ExcludeTags": ["test"] 
	},
	"Video": {
	},
	"Flags": {
	},
	"Export": {
	},
}
```

### CODE REVIEW

**Comment by andybak** [comment]
Eligible is bugging me. How is that better than include/exclude? It was meant to make the fact that one has priority over the other less ambiguous but I'm not sure how it does.

**Comment by insominx** [comment]
> Eligible is bugging me. How is that better than include/exclude? It was meant to make the fact that one has priority over the other less ambiguous but I'm not sure how it does.

That was the intention but after some reflection, I think you're right.  TagInclude and TagExclude are more easily understood and "eligible" doesn't necessarily imply order.  Being more formal about order would convolute this process and probably not add much value.

**Comment by andybak** [comment]
@mikeskydev Any strong opinions on this commit? It's fairly niche as it stands but it has the potential to be a good base for other features.

File: Assets/Scripts/BrushCatalog.cs [hunk 1]
```diff
@@ -181,6 +181,34 @@ public void BeginReload()
...
+        {
+            string[] includeTags = App.UserConfig.Brushes.IncludeTags;
+            string[] excludeTags = App.UserConfig.Brushes.ExcludeTags;
+
+            if (includeTags == null)
+            {
+                Debug.LogError("There will be no brush because none of them are tagged 'include'");
+            }
+
+            // Filter m_GuiBrushList down to those that are both 'eligible' and not 'excluded'
```
Comment (mikeage) [inline Assets/Scripts/BrushCatalog.cs:195]: please replace `'eligible'` with `'included'` to align with the newer name

**Review by mikeage — COMMENTED** [review]
one nitpick about the terminology (and it's just in a comment!) but conceptually, this looks reasonable to me.

File: Assets/Scripts/BrushCatalog.cs (hunk 1 above)
Comment (insominx) [reply Assets/Scripts/BrushCatalog.cs]: ✔

**Comment by mikeskydev** [comment]
> @mikeskydev Any strong opinions on this commit? It's fairly niche as it stands but it has the potential to be a good base for other features.

I think this is a great idea! when user brushes get merged I can imagine this panel getting quite busy, and this is a good base to start the categorization process!

**Comment by andybak** [comment]
1. Other than doing their own custom build - is there any way for the end user to assign tags?
2. What tags have been assigned in this PR?

**Comment by insominx** [comment]
> 1. Other than doing their own custom build - is there any way for the end user to assign tags?

Sadly, no.  Users would need to know a reliable string identifier and currently the main thing they would see is the ``BrushDescriptor.Description``.  Is there a better way?

> 2. What tags have been assigned in this PR?

A group of brushes that cater to whiteboarding have been given the additional tag "classroom".

**Comment by andybak** [comment]
> 
> 
> > 1. Other than doing their own custom build - is there any way for the end user to assign tags?
> 
> Sadly, no. Users would need to know a reliable string identifier and currently the main thing they would see is the `BrushDescriptor.Description`. Is there a better way?

Yeah. I faced the same dilemma with the API where you can change brushes by name. There is a unique descriptor but it doesn't match what the user sees in the UI. I was originally using Durable Name which is guaranteed to be unique but decided to switch to "Description" which is currently unique. I'm still not entirely happy about this but in my case it seemed to be acceptable from a pragmatic standpoint. 

I think we need to either:

1. Stick with durable name but document the mapping somewhere
2. Be liberal in what we accept. Maybe have some common code to do a best effort mapping (1. Try to match DurableName 2. Strip special characters and try to match Description. 3. Try to match Description some other way) For the API I quite like the latter option as it's meant to be playful and approachable rather than watertight and rigorous.

What do you think? This could be punted to a future PR but would make this PR a lot more useful.
 
> > 1. What tags have been assigned in this PR?
> 
> A group of brushes that cater to whiteboarding have been given the additional tag "classroom".

Just to give this PR a bit more "meat" could we get a few other tags in? Maybe original, experimental, particle, tube, additive, hull? Happy to help with this.

In summary:

1. I'm happy with this PR as it stands.
2. It would be great to add a few more tags but it's not a blocker.
3. Future PRs could tackle allowing users to set their own tags. This would entail making a decision about string to brush mapping.

**Comment by insominx** [comment]
> I think we need to either:
> 
> 1. Stick with durable name but document the mapping somewhere
> 2. Be liberal in what we accept. Maybe have some common code to do a best effort mapping (1. Try to match DurableName 2. Strip special characters and try to match Description. 3. Try to match Description some other way) For the API I quite like the latter option as it's meant to be playful and approachable rather than watertight and rigorous.
> 
> What do you think? This could be punted to a future PR but would make this PR a lot more useful.

"Best effort mapping" sounds like a good approach.  I'll take a crack at a it.  Stay tuned!

> Just to give this PR a bit more "meat" could we get a few other tags in? Maybe original, experimental, particle, tube, additive, hull? Happy to help with this.
> 

Another good idea.  Please do jump in and tag!  I don't want anyone asking "where's the beef"?

**Comment by insominx** [comment]
@andybak I've made some additions based on your recommendation.  Let me know what you think.
